### PR TITLE
[android] Enable several C++ Interop and other tests

### PR DESCRIPTION
**Pull request by finagolfin**
Also, update NDK and Ubuntu versions to the latest in the Android doc, enable inline bridging when building the compiler for Android, and update the default arch set by `build-script` to AArch64, as that is by far the most commonly used now.

These C++ Interop tests were all disabled over the years because of build errors with the Android NDK, but @hyp's recent pull #72161 adding a modularized Bionic module map fixed them and the inline bridging issue in the compiler itself. Alex, let me know if these tests all pass for you too.

@drodriguez, perhaps you'd like to take a look at this pull too.

### CODE REVIEW

File: test/Interop/SwiftToCxx/core/unsigned-return-type-no-zext.cpp [hunk 1]
```diff
@@ -4,7 +4,6 @@
 // RUN: %FileCheck %s < %t/ir.ll
 
 // UNSUPPORTED: OS=windows-msvc
-// XFAIL: OS=linux-android, OS=linux-androideabi
```
Comment (finagolfin) [inline test/Interop/SwiftToCxx/core/unsigned-return-type-no-zext.cpp:7]: I disabled this last year, 12f20ec75, because it was failing with the clang 14 in NDK 25 at that time, but it builds fine with clang 17 in NDK 26 now.

File: test/Interpreter/llvm_link_time_opt.swift [hunk 2]
```diff
@@ -1,8 +1,6 @@
 // UNSUPPORTED: OS=windows-msvc
 // static library is not well supported yet on Windows
 
-// XFAIL: OS=linux-android, OS=linux-androideabi
```
Comment (finagolfin) [inline test/Interpreter/llvm_link_time_opt.swift:4]: Similarly, [I disabled this last year because it didn't work with clang 14 in NDK 25](https://github.com/swiftlang/swift/pull/65968/files#r1273589720), but it works with clang 17 in NDK 26 now.

File: test/lit.cfg [hunk 3]
```diff
@@ -615,7 +615,7 @@ else:
             config.swift_driver_test_options,
         )
     )
-    toolchain_lib_dir = make_path(config.swift_lib_dir, 'swift', host_os)
+    toolchain_lib_dir = make_path(config.swift_lib_dir, 'swift', 'android' if kIsAndroid else host_os)
```
Comment (finagolfin) [inline test/lit.cfg:618]: This path was [just added late last year](https://github.com/swiftlang/swift/commit/9c9010e5b750#diff-a041dc49aeef2cbc59d598b3df453d06221ec2704c2c076590d5835cbb77de44R557) and broke building several macro tests natively on Android that have started using it.

File: test/lit.cfg [hunk 4]
```diff
@@ -2428,8 +2428,12 @@ elif not kIsWindows and not run_os == 'wasi':
         lit_config.note('Testing with the just-built libraries')
 
     lit_config.note('Library load path: {0}'.format(os.path.pathsep.join(target_stdlib_path)))
+    env_path = "/usr/bin/env "
+    # Android doesn't have /usr/bin/, so just grab whatever `env` is in the PATH.
```
Comment (compnerd) [inline test/lit.cfg:2432]: It does have /bin/env which is pretty standard, why not use that?

File: utils/build_swift/build_swift/driver_arguments.py [hunk 5]
```diff
@@ -1403,7 +1403,7 @@ def create_argument_parser():
 
     option('--android-arch', store,
            choices=['armv7', 'aarch64', 'x86_64'],
-           default='armv7',
+           default='aarch64',
```
Comment (compnerd) [inline utils/build_swift/build_swift/driver_arguments.py:1406]: This seems unrelated? It's a welcome change, just seems not related to the test enablement.

Comment (finagolfin) [reply utils/build_swift/build_swift/driver_arguments.py:1406]: Unrelated to the title, yes, but these are all Android changes, and I specifically mention these non-test changes in the commit message.

Comment (compnerd) [reply utils/build_swift/build_swift/driver_arguments.py:1406]: Sure, but it still helps to keep changes focused :)

File: test/lit.cfg (hunk 3 above)
Comment (compnerd) [reply test/lit.cfg:618]: Hmm, the `host_os` should be android if we are building for android right?

**Comment by compnerd** [comment]
@swift-ci please smoke test

File: utils/build_swift/build_swift/driver_arguments.py (hunk 5 above)
Comment (egorzhdan) [reply utils/build_swift/build_swift/driver_arguments.py:1406]: Right, this should probably be a separate patch, since it isn't just a test-only change.

File: test/lit.cfg (hunk 3 above)
Comment (finagolfin) [reply test/lit.cfg:618]: [`host_os` is parsed from the host triple](https://github.com/swiftlang/swift/blob/b13d5607d1720c276357518b99cbce7f9ad974ca/test/lit.cfg#L418), which on Android is `linux`. That heuristic works on most platforms, but not for an Android host.

File: test/lit.cfg (hunk 4 above)
Comment (finagolfin) [reply test/lit.cfg:2432]: I'll try running the tests with that instead.

File: test/lit.cfg (hunk 3 above)
Comment (compnerd) [reply test/lit.cfg:618]: Ah, that is annoying. Thanks for the explanation!

File: utils/build_swift/build_swift/driver_arguments.py (hunk 5 above)
Comment (finagolfin) [reply utils/build_swift/build_swift/driver_arguments.py:1406]: > Right, this should probably be a separate patch, since it isn't just a test-only change.

It never says it's a test-only change, the label used is `[android]` and the commit message lists the non-test changes.

If you would like, I can split this into separate commits, but I see no reason to go farther.

**Comment by finagolfin** [comment]
Sigh, `File "/home/build-user/swift/test/lit.cfg", line 618, in <module>
    toolchain_lib_dir = make_path(config.swift_lib_dir, 'swift', 'android' if kIsAndroid else host_os)
NameError: name 'kIsAndroid' is not defined`.

I had simply hacked in `make_path(config.swift_lib_dir, 'swift', 'android')` in my local testing and then cleaned it up with the `kIsAndroid` check before submitting, without checking if that variable had been defined yet.

I'll clean that up and run the validation suite natively on Android on the entirety of this pull before asking for another CI run.

File: utils/build_swift/build_swift/driver_arguments.py (hunk 5 above)
Comment (egorzhdan) [reply utils/build_swift/build_swift/driver_arguments.py:1406]: The PR title doesn't mention any non-test changes, which makes it confusing. I initially assumed that this is a test-only change.
I think the arch change should be its own commit.

Comment (finagolfin) [reply utils/build_swift/build_swift/driver_arguments.py:1406]: OK, I will split this commit up.

File: test/lit.cfg (hunk 4 above)
Comment (finagolfin) [reply test/lit.cfg]: Done

File: utils/build_swift/build_swift/driver_arguments.py (hunk 5 above)
Comment (finagolfin) [reply utils/build_swift/build_swift/driver_arguments.py:1406]: Done

**Comment by finagolfin** [comment]
Alright, fixed the `lit.cfg` error and ran this entire pull through the compiler validation suite natively on Android, before splitting it up into narrower commits and rebasing it here.

This is ready for review and other CI run.

File: test/lit.cfg [hunk 6]
```diff
@@ -417,6 +421,11 @@ if run_os.startswith('wasi'):
 # Parse the host triple
 (host_cpu, host_vendor, host_os, host_vers) = re.match('([^-]+)-([^-]+)-([^0-9-]+)(.*)', config.host_triple).groups()
 
+# Manually set the host OS on Android, because the host OS from the triple says
+# `linux`.
+if kIsAndroid:
+    host_os = 'android'
```
Comment (finagolfin) [inline test/lit.cfg:427]: Rather than overriding this just for the new `toolchain_lib_dir` below, I overrode `host_os` itself on Android, as it is only used for that and `target-same-as-host` just below, which is only checked by a single macOS-only test, so changing this for Android hosts should be fine.

File: utils/build_swift/tests/expected_options.py [hunk 7]
```diff
@@ -768,7 +768,7 @@ class BuildScriptImplOption(_BaseOption):
     ChoicesOption('--swift-analyze-code-coverage',
                   choices=['false', 'not-merged', 'merged']),
     ChoicesOption('--android-arch',
-                  choices=['armv7', 'aarch64']),
+                  choices=['armv7', 'aarch64', 'x86_64']),
```
Comment (finagolfin) [inline utils/build_swift/tests/expected_options.py:771]: This never triggered before because this just checks if it's a subset, so updating to all three arches supported now.

**Comment by finagolfin** [comment]
@kateinoigakukun, need a CI run here.

@egorzhdan, maybe you can review.

**Comment by finagolfin** [comment]
@rauhul, need a CI run here.

**Comment by finagolfin** [comment]
@bnbarham, ready for a CI run here.

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by finagolfin** [comment]
Alright, passed CI, just need @egorzhdan or some one to review.

**Review by egorzhdan — APPROVED** [review]
C++ interop changes LGTM

**Comment by finagolfin** [comment]
@etcwilde, if you would review the few remaining Android changes, we can get this in.

**Review by etcwilde — APPROVED** [review]

**Comment by finagolfin** [comment]
@bnbarham, ready for merge.

**Comment by finagolfin** [comment]
@hyp, would you merge?

**Comment by bnbarham** [comment]
Sorry, missed initial ping. Merged!
